Library/Homebrew/formula.rb

```diff
@@ -476,6 +476,29 @@ def compiler_failures
     active_spec.compiler_failures
   end
 
+  # The best spec we can upgrade the {Formula} to so it can satisfy dependency.
+  def best_upgrade_spec_sym
```

Comment:
Ok, I understand this a bit more. I think it should probably live on `Dependency` and be called something like `def oldest_spec_for_dependency_upgrade`

Maybe add this as a comment above this method (if I've understood correctly)
- If the dependency is installed as a `stable` version but outdated: `brew upgrade` to the latest stable version
- If the dependency is installed as a `devel` version but outdated:
  - if the latest `stable` version is newer than the installed `devel` version: `brew upgrade` to the latest stable version
  - if the latest `devel` version is newer than the installed `devel` version and latest `stable` version: `brew upgrade` to the latest devel version

I think HEAD shouldn't ever be upgraded to `stable`/`devel` or perhaps even the latest `HEAD` on `install` or `upgrade` of a reverse dependency.

I also think linking doesn't matter but opt-linking does.